connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java

```diff
@@ -0,0 +1,189 @@
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import lombok.Getter;
+import lombok.Setter;
+import org.geysermc.connector.entity.player.SessionPlayerEntity;
+import org.geysermc.connector.network.session.GeyserSession;
+import org.geysermc.connector.network.translators.collision.BoundingBox;
+import org.geysermc.connector.network.translators.world.block.entity.PistonBlockEntity;
+import org.geysermc.connector.utils.Axis;
+
+import java.util.Map;
+
+public class PistonCache {
+    private final GeyserSession session;
+
+    /**
+     * Maps the position of a piston to its block entity
+     */
+    @Getter
+    private final Map<Vector3i, PistonBlockEntity> pistons = new Object2ObjectOpenHashMap<>();
+
+    /**
+     * Maps the position of a moving block to the piston moving it
+     * Positions in this map represent the starting position of the block
+     */
+    @Getter
+    private final Map<Vector3i, PistonBlockEntity> movingBlocksMap = new Object2ObjectOpenHashMap<>();
+
+    @Getter
+    private Vector3d playerDisplacement = Vector3d.ZERO;
+
+    @Getter @Setter
+    private Vector3f playerMotion = Vector3f.ZERO;
+
+    /**
+     * Stores whether a player has/will collide with any moving blocks.
+     */
+    @Getter @Setter
+    private boolean playerCollided = false;
+
+    /**
+     * Stores whether a player has/will collide with any slime blocks.
+     * This is used to prevent movement from being corrected when players
+     * are about to hit a slime block.
+     */
+    @Getter @Setter
+    private boolean playerSlimeCollision = false;
+
+    /**
+     * Stores whether a player is standing on a honey block.
+     * This is used to ignore movement from Bedrock to prevent them from
+     * falling off.
+     */
+    @Getter @Setter
+    private boolean playerAttachedToHoney = false;
+
+    public PistonCache(GeyserSession session) {
+        this.session = session;
+    }
+
+    public void tick() {
+        resetPlayerMovement();
+        if (!pistons.isEmpty()) {
+            pistons.values().forEach(PistonBlockEntity::updateMovement);
+            sendPlayerMovement();
+            sendPlayerMotion();
+            // Update blocks after movement, so that players don't get stuck inside blocks
+            pistons.values().forEach(PistonBlockEntity::updateBlocks);
+
+            pistons.entrySet().removeIf((entry) -> entry.getValue().canBeRemoved());
+
+            if (pistons.isEmpty() && !movingBlocksMap.isEmpty()) {
+                session.getConnector().getLogger().error("The moving block map has de-synced!");
+                for (Map.Entry<Vector3i, PistonBlockEntity> entry : movingBlocksMap.entrySet()) {
+                    Vector3i position = entry.getKey();
+                    PistonBlockEntity pistonBlockEntity = entry.getValue();
+                    session.getConnector().getLogger().error("Moving Block at " + position + " was previously owned by the piston at " + pistonBlockEntity.getPosition());
```

Comment:
The `position` and `pistonBlockEntity` values can be inlined as they are only referenced once inside the log method.